libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift

```diff
@@ -0,0 +1,328 @@
+    }
+    
+    public override func rightMouseDown(with event: NSEvent) {
+        let local = viewCoordinates(event)
+        mouse_button(wsHandle, Float(local.x), Float(local.y), true, false, event.modifierFlags.contains(.shift), event.modifierFlags.contains(.control), event.modifierFlags.contains(.option), event.modifierFlags.contains(.command))
+        setNeedsDisplay(self.frame)
+    }
+    
+    public override func rightMouseUp(with event: NSEvent) {
+        let local = viewCoordinates(event)
+        mouse_button(wsHandle, Float(local.x), Float(local.y), false, false, event.modifierFlags.contains(.shift), event.modifierFlags.contains(.control), event.modifierFlags.contains(.option), event.modifierFlags.contains(.command))
+        setNeedsDisplay(self.frame)
+    }
+    
+    public override func scrollWheel(with event: NSEvent) {
+        scroll_wheel(wsHandle, Float(event.scrollingDeltaY))
+        setNeedsDisplay(self.frame)
+    }
+    
+    public override func keyDown(with event: NSEvent) {
+        setClipboard()
+        if pasteImageInClipboard(event) {
+            return
+        }
+        
+        key_event(wsHandle, event.keyCode, event.modifierFlags.contains(.shift), event.modifierFlags.contains(.control), event.modifierFlags.contains(.option), event.modifierFlags.contains(.command), true, event.characters)
+        setNeedsDisplay(self.frame)
+    }
+    
+    public override func viewDidChangeEffectiveAppearance() {
+        dark_mode(wsHandle, isDarkMode())
+        setNeedsDisplay(self.frame)
+    }
+    
+    // https://stackoverflow.com/a/53218688/1060955
+    func isDarkMode() -> Bool {
+        self.effectiveAppearance.bestMatch(from: [.darkAqua, .aqua]) == .darkAqua
+    }
+    
+    public override func keyUp(with event: NSEvent) {
+        key_event(wsHandle, event.keyCode, event.modifierFlags.contains(.shift), event.modifierFlags.contains(.control), event.modifierFlags.contains(.option), event.modifierFlags.contains(.command), false, event.characters)
+        setNeedsDisplay(self.frame)
+    }
+    
+    func setClipboard(){
+        let pasteboardString: String? = NSPasteboard.general.string(forType: .string)
+        if let theString = pasteboardString {
+            system_clipboard_changed(wsHandle, theString)
+        }
+        self.pasteBoardEventId = NSPasteboard.general.changeCount
+    }
+    
+    func importFromPasteboard(_ pasteBoard: NSPasteboard, isDropOperation: Bool) -> Bool {
+        if let data = pasteBoard.data(forType: .png) ?? pasteBoard.data(forType: .tiff),
+           let url = createTempDir() {
+            let imageUrl = url.appendingPathComponent(String(UUID().uuidString.prefix(10).lowercased()), conformingTo: .png)
+            
+            do {
+                try data.write(to: imageUrl)
+            } catch {
+                return false
+            }
+            
+            //            if let lbImageURL = workspaceState!.importFile(imageUrl) {
```

Comment:
what's happening here @smailbarkouch?

Reply:
Weird, I already fixed all of paste import. I think someone forced pushed it out of the history, probably when we were trying to fix merge conflicts.